### ISSUE TITLE
Make it easier to generate package-info and constructor annotations with the JavaGenerator

### ISSUE THREAD

**Issue by agentgt** [jOOQ/jOOQ#13845]
### Use case:

As for as I can tell it is not easy to generate your own custom `package-info.java` even if you extend the Java Generators.

It is also not easy to put annotations on the record constructors. If you like I can file that as a separate issue.

### Possible solution you'd like to see:

Ideally I would like to be able to have methods on the JavaGenerator that I can override that get called when a package is created. Perhaps the method passes what type of package (e.g. table, record, pojo).

### Possible workarounds:

This workaround generates a `package-info.java` for Eclipse null analysis as well as a custom annotation on the record constructor that takes all the arguments called `Default`. The latter is for MapStruct so that it knows which constructor to use as well as some other libraries we use.

Normally I would be fine with a workaround like this but it relies on using a ThreadLocal and the fact that `generateConstructorPropertiesAnnotationOnRecords` predicate is called exactly before the constructor generation method is called (which I believe is `generateRecordConstructor` which is private).

I guess one solution to the above would be to make the constructor properties annotation generation not private.

Anyway my hack is below.

```java
public class CustomJavaGenerator extends JavaGenerator {

    private final ThreadLocal<@Nullable JavaWriter> currentJavaWriter = new ThreadLocal<>();
    private final ConcurrentHashMap<String, String> packageMap = new ConcurrentHashMap<>();

    @Override
    protected void generateRecord(
            TableDefinition table) {
        String packageName = getStrategy().getJavaPackageName(table, Mode.RECORD);
        File recordFile = getFile(table, Mode.RECORD);
        if (packageName != null && packageMap.putIfAbsent(packageName, packageName) == null) {
            File packageDirectory = recordFile.getParentFile();
            File packageInfoFile = new File(packageDirectory, "package-info.java");
            JavaWriter packageInfo = newJavaWriter(packageInfoFile);
            packageInfo.println("@" + "org.eclipse.jdt.annotation.NonNullByDefault");
            packageInfo.println("package " + packageName + ";");
            closeJavaWriter(packageInfo);
        }
        JavaWriter out = newJavaWriter(getFile(table, Mode.RECORD));

        if (currentJavaWriter.get() != null) {
            log.error("Java writer wasn't cleared from thread local");
        }
        currentJavaWriter.set(out);
        try {
            log.info(
                    "Generating record",
                    out.file()
                        .getName());
            generateRecord(table, out);
        }
        finally {
            currentJavaWriter.remove();
        }
        closeJavaWriter(out);
    }

    @Override
    public boolean generateConstructorPropertiesAnnotationOnRecords() {
        JavaWriter out = currentJavaWriter.get();
        // if out is not null then we know we are generating a record
        if (out != null) {
            out.println("@%s", Default.class);
        }

        return super.generateConstructorPropertiesAnnotationOnRecords();
    }

    private static final JooqLogger log = JooqLogger.getLogger(CustomJavaGenerator.class);

}
```

Notice the above `package-info` only works with records. I can't recall why had trouble with generating `package-info` with tables and pojos using the same approach. When I find out I'll update the bug. The workaround for those guys is just to checkin the package-info in the right package and have the build overlay it on the code generated package (which automatically happens for gradle/maven if you just put it in `src/main/java`).

### Versions:

- jOOQ: 3.17.2
- Java: 17

**Comment by lukaseder** [comment]
Thanks for your feature request.

> If you like I can file that as a separate issue.

Yes please. One thing may be simple to implement whereas the other may be tricky and controversial. There's not much gained from mixing concerns in an issue, whereas keeping things separate has only benefits.

**Comment by agentgt** [comment]
Yeah I got a little lazy and could not decide if the package-info was even worth the complaint given that one is actually easily solvable and has workarounds.

I will file another issue.

**Comment by lukaseder** [comment]
> given that one is actually easily solvable and has workarounds.

While workarounds may be easy, nothing in product development is ever easy... In the case of package infos, the generator strategy allows for generating arbitrary package structures, including custom ones where each table is in a separate package. What would happen to package info files in such a case?

In your particular case, there's again the question whether packages are interpreted as hierarchical or not for the sake of annotation processing (related discussion: https://github.com/jOOQ/jOOQ/issues/9115). While I don't agree packages should be considered hierarchical (because the Java language doesn't see them this way), some users will still want it.

In any case, you could probably just use any other maven or gradle plugin to generate the package-info file after jOOQ generates its code? Or you could add a new `src/other/java` directory that contains the package-info file in the same package (split packages are still allowed post Java 9 in the same module from different source directories, if I'm not mistaken?)

In any case, while I can see some benefit of generating `package-info.java` files, this feature needs to get many more popularity votes before I'll endeavour adding it. The workarounds are certainly much more viable for the time being.

**Comment by agentgt** [comment]
I agree. I was hesitant to file either bug but I figured someone might find the code snippet useful.

To be honest it is a testament to the design of the generator that I was still able to find workarounds for both cases!

Thanks for the thoughtful reply and I hope I didn’t take up too much time.

**Comment by lukaseder** [comment]
No worries. Always good to hear new ideas and collect feedback from users.
